### PR TITLE
Jk moe

### PR DESCRIPTION
**Pull request by kuba-krj**
Adding MoE to our codebase, written with the assistance of Codex.

### MFU
MFU calculated on 1 GPU is not great: ~8% with the following settings:
```
  dmodel: 1024
  dff: 2816
  dhead: 64
  n_blocks: 16
  q_heads: 16
  kv_heads: 16

  num_experts: 16
  num_experts_per_tok: 1
  capacity_factor: 1.25
```
with `batch size=32` (the largest that could fit on 1 GPU), `seq_len=1024`. This is ~2x slower than dense with the same number of active params, trained with `batch size=64` (also the largest that could fit). Possibly MFU is better on multi-gpu due to a larger batch size that we can use, but the exps are waiting in the queue, I will update when they are finished.

### Correctness
I compared dense model (settings as above) with MoE where E={1, 2, 4, 16}. The results look reasonably - E=1 matches dense, and models get better with more experts.

Link to verification experiments: [wandb project](https://wandb.ai/ideas_cv/moe%20PR?nw=nwuserkrjjakub)

### CODE REVIEW

File: configs/_cluster/entropy.yaml
Comment (j321m) [inline configs/_cluster/entropy.yaml:1]: remove this file from PR

File: configs/moe_example_run.yaml
Comment (j321m) [inline configs/moe_example_run.yaml:1]: remove / rename

File: src/core/moe.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        dmodel: int,
+        dff: int,
+        num_experts: int,
+        num_experts_per_tok: int,
+        capacity_factor: float = 1.25,
+        moe_load_balancing_loss_factor: float = 0.0,
+        moe_router_z_loss_factor: float = 0.0,
+        activation_function: str = "swiglu",
+        init_scale: float = 1.0,
+        **_ignored_kwargs,
```
Comment (j321m) [inline src/core/moe.py]: is `**_ignored_kwargs` necessary?

File: src/core/moe.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        # Router
+        router_logits = torch.einsum(
+            "th,eh->te",
+            hidden_states,
+            self.router_weight,
+        )
+        router_logits = router_logits.to(dtype=torch.float32)
+        router_probs = F.softmax(router_logits, dim=-1)
+        # For each token, keep only the top-k experts and their routing probabilities
+        topk_probs, selected_experts = torch.topk(
```
Comment (j321m) [inline src/core/moe.py:79]: question: should the routing weights sum to 1, when `num_experts_per_tok` > 1?

File: src/core/trainer.py [hunk 3]
```diff
@@ -228,11 +312,19 @@ def _update_processed_tokens(self, batch):
 
     def log_metrics(self, loss, grad_norm):
```
Comment (j321m) [inline src/core/trainer.py]: function `log_metrics` ignores the loss argument. also i'm not sure if i like `self._last_reported_loss`, `self._last_moe_router_z_loss`, ect., it makes the code more errorprone in my opinion but I'm open for discussion

File: src/core/trainer.py [hunk 4]
```diff
@@ -228,11 +312,19 @@ def _update_processed_tokens(self, batch):
...
         self.metric_logger.set_tokens(self.processed_tokens)
-        self.metric_logger.log("train/loss", loss.item())
+        self.metric_logger.log("train/loss", self._last_reported_loss.item())
+        self.metric_logger.log(
```
Comment (j321m) [inline src/core/trainer.py]: MoE metrics will get logged even for dense models, do we want that?

File: src/core/trainer.py
Comment (j321m) [inline src/core/trainer.py]: function `eval` calls `self.calculate_loss(batch)` which overwrites `self._last_reported_loss`, resulting in same eval and train loss.

It works now, since `log_metrics` gets called after `eval`

File: src/core/moe.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+        self.dmodel = dmodel
+        self.dff = dff
+        self.num_experts = num_experts
+        self.num_experts_per_tok = num_experts_per_tok
+        self.capacity_factor = capacity_factor
+        self.moe_load_balancing_loss_factor = moe_load_balancing_loss_factor
+        self.moe_router_z_loss_factor = moe_router_z_loss_factor
+        self.is_moe = True
+        self.aux_loss = None
```
Comment (j321m) [inline src/core/moe.py]: what is `aux_loss` for? it is set to the same value as `load_balancing_loss`

**Review by j321m — COMMENTED** [review]
claude assisted review.

I'd like to see multigpu run (smoke test)

please address the comments

**Comment by kuba-krj** [comment]
> I'd like to see multigpu run (smoke test)

Link to multrigpu run: [wandb](https://wandb.ai/ideas_cv/moe%20PR/runs/p6ia98n8?nw=nwuserkrjjakub)

File: configs/moe_example_run.yaml
Comment (kuba-krj) [reply configs/moe_example_run.yaml:1]: Renamed

File: src/core/moe.py (hunk 1 above)
Comment (kuba-krj) [reply src/core/moe.py]: As far as I understand our config system, it's necessary to be able to keep MoE configs similar to how it is done in small_moe.yaml, where we just set 
```
ff_layer_fn:
        _target_: src.core.moe.MoE
```
to use MoE (because we keep ff_layer_fn from base config and only replace _ target _ in it). Please let me know if you prefer to change the config structure to sth like - override /ff_layer@model.encoder.block_fn.ff_layer_fn: moe - we can then get rid of **_ignored_kwargs

File: src/core/moe.py (hunk 5 above)
Comment (kuba-krj) [reply src/core/moe.py]: removed

File: src/core/trainer.py (hunk 4 above)
Comment (kuba-krj) [reply src/core/trainer.py]: removed moe metric logging for dense

File: src/core/trainer.py (hunk 3 above)
Comment (kuba-krj) [reply src/core/trainer.py]: Refactored

File: src/core/trainer.py
Comment (kuba-krj) [reply src/core/trainer.py]: Refactored

File: src/core/moe.py (hunk 2 above)
Comment (kuba-krj) [reply src/core/moe.py:79]: I added the option to normalize

**Review by kuba-krj — COMMENTED** [review]
Added changes and ran a test run identical to the previous 2-gpu to see if the results are unchanged: [wandb link](https://wandb.ai/ideas_cv/moe%20PR/runs/rssexqk1?nw=nwuserkrjjakub). Pls let me know if the PR looks good now or if additional changes are needed

File: src/core/moe.py (hunk 1 above)
Comment (j321m) [reply src/core/moe.py]: I think it's better to have separate base yamls for dense and moe, than **kwargs.

**Comment by j321m** [comment]
removing **kwargs is very important, separate dense and MoE config lines should solve the problem.

File: src/core/moe.py (hunk 1 above)
Comment (j321m) [reply src/core/moe.py]: override /[ff_layer@model.encoder.block_fn.ff_layer_fn](mailto:ff_layer@model.encoder.block_fn.ff_layer_fn): moe
is also a good idea (even better, but may need some config refactoring)

**Review by j321m — APPROVED** [review]
awesome
